internal/core/data/v2/application/event.go

```diff
@@ -0,0 +1,91 @@
+//
+// Copyright (C) 2020 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package application
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	dataContainer "github.com/edgexfoundry/edgex-go/internal/core/data/container"
+	v2DataContainer "github.com/edgexfoundry/edgex-go/internal/core/data/v2/bootstrap/container"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/correlation"
+
+	"github.com/edgexfoundry/go-mod-bootstrap/bootstrap/container"
+	"github.com/edgexfoundry/go-mod-bootstrap/di"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/clients"
+	model "github.com/edgexfoundry/go-mod-core-contracts/v2/models"
+
+	msgTypes "github.com/edgexfoundry/go-mod-messaging/pkg/types"
+)
+
+// The AddEvent function accepts the new event model from the controller functions
+// and invokes addEvent function in the infrastructure layer
+func AddEvent(e model.Event, ctx context.Context, dic *di.Container) (string, error) {
+	configuration := dataContainer.ConfigurationFrom(dic.Get)
+	dbClient := v2DataContainer.DBClientFrom(dic.Get)
+	lc := container.LoggingClientFrom(dic.Get)
+
+	err := checkDevice(e.DeviceName, ctx, dic)
+	if err != nil {
+		return "", err
+	}
+
+	// Add the event and readings to the database
+	if configuration.Writable.PersistData {
+		correlationId := correlation.FromContext(ctx)
+		e.CorrelationId = correlationId
+		addedEvent, err := dbClient.AddEvent(e)
+		if err != nil {
+			return "", err
+		}
+		e = addedEvent
+
+		lc.Info(fmt.Sprintf(
+			"Event created on DB successfully. Event-id: %s, Correlation-id: %s ",
+			e.Id,
+			correlationId,
+		))
+	}
+
+	putEventOnQueue(e, ctx, dic) // Push event to message bus for App Services to consume
+
+	return e.Id, nil
+}
+
+// Put event on the message queue to be processed by the rules engine
+func putEventOnQueue(evt model.Event, ctx context.Context, dic *di.Container) {
```

Comment:
This needs to publish the Event DTO, not Event model, to the message bus. The DTO will have the API version from https://github.com/edgexfoundry/go-mod-core-contracts/pull/255

Reply:
Will update this until https://github.com/edgexfoundry/go-mod-core-contracts/pull/255 is merged.

Reply:
Updated with Event DTO to publish on the message bus.